### PR TITLE
feat: delitem for DocumentArray

### PR DESCRIPTION
**Pull request by JoanFM**
Signed-off-by: Joan Fontanals Martinez <joan.martinez@jina.ai>

Goals:
Implement `__del_item__` for DocumentArray and DocumentArrayStacked

### CODE REVIEW

**Review by samsja — CHANGES_REQUESTED** [review]
regarding DocumentArrayStackedI think we should not even expose delitem at the stack level. The philosophy so far was that if users need to do smth slow: "append, ..." then they need to first unstack.

We maybe can do some optimization later to allow such operation on the DocumentArrayStacked but I would consider this for later

**Comment by JoanFM** [comment]
> regarding DocumentArrayStackedI think we should not even expose delitem at the stack level. The philosophy so far was that if users need to do smth slow: "append, ..." then they need to first unstack.
> 
> We maybe can do some optimization later to allow such operation on the DocumentArrayStacked but I would consider this for later

@JohannesMessner what do you think? should I just remove it for `DocumentArrayStacked`?

File: docarray/array/array.py [hunk 1]
```diff
@@ -381,3 +384,22 @@ def traverse_flat(
...
+
+    @overload
+    def __delitem__(self: T, key: IndexIterType) -> None:
+        ...
+
+    def __delitem__(self, key) -> None:
+        key = self._normalize_index_item(key)
+
+        if key is None:
+            return
```
Comment (samsja) [inline docarray/array/array.py:400]: what is happening when you do `return` alone ?

File: docarray/array/array.py [hunk 2]
```diff
@@ -381,3 +384,22 @@ def traverse_flat(
...
+        if type(key) == slice:
+            del self._data[key]
+
+        if isinstance(key, int):
+            del self._data[key]
```
Comment (samsja) [inline docarray/array/array.py:405]: not sure but cannot we just do 
```python
del self._data[key]
```
I this the if statement do nothing

File: docarray/array/array.py (hunk 1 above)
Comment (JoanFM) [reply docarray/array/array.py:400]: returns None

File: docarray/array/array.py (hunk 2 above)
Comment (JoanFM) [reply docarray/array/array.py:405]: makes sense

File: docarray/array/array_stacked.py [hunk 3]
```diff
@@ -49,10 +48,14 @@ class DocumentArrayStacked(AnyDocumentArray):
...
     you can be precise a Document schema by using the `DocumentArray[MyDocument]`
-    syntax where MyDocument is a Document class  (i.e. schema).
+    syntax where MyDocument is a Document class (i.e. schema).
     This creates a DocumentArray that can only contains Documents of
     the type 'MyDocument'.
 
+    .. note::
+        Deleting elements of a DocumentArrayStacked will rebuild the tensors from a
+        reference to the updated unstacked DocumentArray. This operation may be slow
+         as this class is not designed for so regular update.
```
Comment (JohannesMessner) [inline docarray/array/array_stacked.py:58]: This should be gone now, right?

**Review by JohannesMessner — DISMISSED** [review]

File: docarray/array/array_stacked.py (hunk 3 above)
Comment (JoanFM) [reply docarray/array/array_stacked.py:58]: sure

Comment (JoanFM) [reply docarray/array/array_stacked.py]: recheck

**Review by samsja — APPROVED** [review]

**Pull request merged**